pkg/logcli/client/file.go

```diff
@@ -0,0 +1,264 @@
+		Status: loghttp.QueryStatusSuccess,
+		Data:   f.labels,
+	}, nil
+}
+
+func (f *FileClient) ListLabelValues(name string, quiet bool, start, end time.Time) (*loghttp.LabelResponse, error) {
+	i := sort.SearchStrings(f.labels, name)
+	if i < 0 {
+		return &loghttp.LabelResponse{}, nil
+	}
+
+	return &loghttp.LabelResponse{
+		Status: loghttp.QueryStatusSuccess,
+		Data:   []string{f.labelValues[i]},
+	}, nil
+}
+
+func (f *FileClient) Series(matchers []string, start, end time.Time, quiet bool) (*loghttp.SeriesResponse, error) {
+	m := len(f.labels)
+	if m > len(f.labelValues) {
+		m = len(f.labelValues)
+	}
+
+	lbs := make(loghttp.LabelSet)
+	for i := 0; i < m; i++ {
+		lbs[f.labels[i]] = f.labelValues[i]
+	}
+
+	return &loghttp.SeriesResponse{
+		Status: loghttp.QueryStatusSuccess,
+		Data:   []loghttp.LabelSet{lbs},
+	}, nil
+}
+
+func (f *FileClient) LiveTailQueryConn(queryStr string, delayFor time.Duration, limit int, start time.Time, quiet bool) (*websocket.Conn, error) {
+	return nil, fmt.Errorf("LiveTailQuery: %w", ErrNotSupported)
+}
+
+func (f *FileClient) GetOrgID() string {
+	return f.orgID
+}
+
+type limiter struct {
+	n int
+}
+
+func (l *limiter) MaxQuerySeries(userID string) int {
+	return l.n
+}
+
+type querier struct {
+	r      io.Reader
+	labels labels.Labels
+}
+
+func (q *querier) SelectLogs(ctx context.Context, params logql.SelectLogParams) (iter.EntryIterator, error) {
+	expr, err := params.LogSelector()
+	if err != nil {
+		return nil, fmt.Errorf("failed to extract selector for logs: %w", err)
+	}
+	pipeline, err := expr.Pipeline()
+	if err != nil {
+		return nil, fmt.Errorf("failed to extract pipeline for logs: %w", err)
+	}
+	return newFileIterator(ctx, q.r, q.labels, params, pipeline.ForStream(q.labels))
+}
+
+func (q *querier) SelectSamples(ctx context.Context, params logql.SelectSampleParams) (iter.SampleIterator, error) {
+	return nil, fmt.Errorf("Metrics Query: %w", ErrNotSupported)
+}
+
+func newFileIterator(
+	ctx context.Context,
+	r io.Reader,
+	labels labels.Labels,
+	params logql.SelectLogParams,
+	pipeline logqllog.StreamPipeline,
+) (iter.EntryIterator, error) {
+
+	lr := io.LimitReader(r, defaultMaxFileSize)
+	b, err := ioutil.ReadAll(lr)
+	if err != nil {
+		return nil, err
+	}
+	lines := strings.FieldsFunc(string(b), func(r rune) bool {
+		return r == '\n'
+	})
+
+	if len(lines) == 0 {
+		return iter.NoopIterator, nil
+	}
+
+	stream := logproto.Stream{
+		Labels: labels.String(),
+	}
+
+	// reverse all the input lines if direction == FORWARD
+	if params.Direction == logproto.FORWARD {
+		sort.Slice(lines, func(i, j int) bool {
+			return i > j
+		})
+	}
+
+	for _, line := range lines {
+		parsedLine, _, ok := pipeline.ProcessString(line)
+		if !ok {
+			continue
+		}
+		stream.Entries = append(stream.Entries, logproto.Entry{
+			Timestamp: time.Now(),
+			Line:      parsedLine,
+		})
+	}
+
+	return iter.NewHeapIterator(
+		ctx,
+		[]iter.EntryIterator{iter.NewStreamIterator(stream)},
```

Comment:
HeapIterator with only one iterator is useless this is because you forgot the stream created by LogQL. see my previous comment for labels.

Reply:
Good point. Fixed it. Now using stream iterator with parsed labels.!